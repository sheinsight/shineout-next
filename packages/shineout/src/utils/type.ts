export type RequireAWithB<T, A extends keyof T, B extends keyof T> = 
  | (Omit<T, A> & { [K in A]?: never })
  | (T & Required<Pick<T, B>>)
  | (T & Required<Pick<T, A | B>>)