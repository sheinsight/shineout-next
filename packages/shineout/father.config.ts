import { defineConfig } from 'father';
export default defineConfig({
  extends: '../../.fatherrc.ts',
});
