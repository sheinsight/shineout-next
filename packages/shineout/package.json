{
  "name": "shineout",
  "version": "3.0.0",
  "description": "A components library for React",
  "keywords": [],
  "license": "ISC",
  "author": "",
  "sideEffects": false,
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.js",
  "typings": "./dist/cjs/index.d.ts",
  "scripts": {
    "build": "father build",
    "build:watch": "father dev",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "@sheinx/base": "workspace:*",
    "@sheinx/hooks": "workspace:*",
    "@sheinx/shineout-style": "workspace:*"
  },
  "peerDependencies": {
    "classnames": ">=2",
    "immer": ">=10",
    "react": ">=16.8"
  },
  "devDependencies": {
    "jszip": "^3.10.1"
  }
}
