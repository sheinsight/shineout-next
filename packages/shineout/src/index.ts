export { default as Alert } from './alert';
export { default as Button } from './button';
export { default as Checkbox } from './checkbox';
export { default as DatePicker } from './date-picker';
export { default as Dropdown } from './dropdown';
export { default as EditableArea } from './editable-area';
export { default as Form } from './form';
export { default as Icon } from './icon';
export { default as Image } from './image';
export { default as Input } from './input';
export { default as Popover } from './popover';
export { default as Radio } from './radio';
export { default as Rule } from './rule';
export { default as Spin } from './spin';
export { default as Switch } from './switch';
export { default as Tag } from './tag';
export { default as Tabs } from './tabs';
export { default as Textarea } from './textarea';
export { default as Tooltip } from './tooltip';
export { setConfig } from '@sheinx/base';

export * as TYPE from './type';
