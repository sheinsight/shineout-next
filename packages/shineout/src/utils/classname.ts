import classnames from 'classnames';
import { config } from '@sheinx/base';

interface CSSModule {
  [key: string]: string;
}

/**
 * create a new className generate function, add namespace, handle css module
 * @param style - object; for css module
 * @param module - string
 * @param prefix - string, default value is 'shineout'
 * * */
export default (style: CSSModule, module: string, prefix: string = config.prefix) =>
  (...args: any) => {
    const className = classnames(...args);
    if (!className) return '';

    const ns = `${prefix}${module ? `-${module}` : '-'}`;
    let list: (string | number)[] = className
      .split(' ')
      .map((c) => (c === '_' ? ns : `${ns}-${c}`));
    if ((config as any).cssModule) {
      list = list.map((c: any) => style[c] || c);
    }
    return list.join(' ');
  };
