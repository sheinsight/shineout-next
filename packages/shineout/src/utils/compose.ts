export function compose(...funcs: any) {
  if (funcs.length === 0) {
    return (arg: any) => arg;
  }
  const last = funcs[funcs.length - 1];
  const rest = funcs.slice(0, -1);
  return (...args: typeof last) =>
    rest.reduceRight((composed: any, f: any) => f(composed), last(...args));
}
