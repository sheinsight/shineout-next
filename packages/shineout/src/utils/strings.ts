import { util } from "@sheinx/hooks";

const { devUseWarning } = util;

export function capitalize(str: string) {
  if (typeof str !== 'string') {
    devUseWarning.error('str should be a string');
  }
  return str && str[0].toUpperCase() + str.slice(1);
}

export function substitute(str: string | Function, obj: { [x: string]: any }) {
  if (typeof str === 'string') {
    if (str.indexOf('{') < 0) {
      return str;
    }

    return str.replace(/\\?\{([^{}]+)\}/g, (match, name) => {
      if (match.charAt(0) === '\\') {
        return match.slice(1);
      }
      return obj[name] === null || obj[name] === undefined ? '' : obj[name];
    });
  }
  if (typeof str === 'function') {
    let val = str(obj);
    if (val === obj && typeof val === 'object') {
      val = Object.assign({}, obj);
    }
    return val;
  }

  return '';
}

export function removeProtocol(url: string) {
  if (url.indexOf('http') !== 0) return url;
  try {
    const { href, protocol } = new URL(url);
    return href.slice(protocol.length);
  } catch (error) {
    return url;
  }
}

export function getRTLPosition(position: string) {
  if (!position) return position;
  // position.replace('left', 'right').replace('right', 'left')
  if (position.indexOf('left') !== -1) {
    return position.replace('left', 'right');
  }
  if (position.indexOf('right') !== -1) {
    return position.replace('right', 'left');
  }
  return position;
}

export function getDirectionIconName(mode = 'left', double = false) {
  const rtl = true;
  if (mode === 'left') {
    if (rtl) {
      return double ? 'ArrowRightDouble' : 'AngleRight';
    }
    return double ? 'ArrowLeftDouble' : 'AngleLeft';
  }

  if (rtl) {
    return double ? 'ArrowLeftDouble' : 'AngleLeft';
  }
  return double ? 'ArrowRightDouble' : 'AngleRight';
}
