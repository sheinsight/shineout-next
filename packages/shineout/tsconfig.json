{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "declaration": true,
    "baseUrl": "./",
    "paths": {
      "shineout": ["./src/index"],
      "@sheinx/base": ["../base/src/index"],
      "@sheinx/hooks": ["../hooks/src/index"]
    }
  },
  "include": ["src"]
}
