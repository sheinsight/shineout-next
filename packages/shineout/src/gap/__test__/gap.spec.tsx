import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import Gap from '..';
import { Button } from 'shineout';
import mountTest from '../../tests/mountTest';
import { displayTest, snapshotTest, styleTest } from '../../tests/utils';
import GapBase from '../__example__/01-base';
import GapCustom from '../__example__/02-custom';
import GapVertical from '../__example__/03-vertical';
import GapAlign from '../__example__/04-align';
import GapSplit from '../__example__/05-split';

const GapDemo = (props: any) => (
  <Gap style={{ width: 400 }} {...props}>
    {Array.from({ length: 4 }).map((_, i) => (
      <Button key={i} type='primary'>
        Item
        {i}
      </Button>
    ))}
  </Gap>
);

const getStyle = (v: number) => `margin-bottom: ${v}px; margin-right: ${v}px;`;
const baseStyle = getStyle(-8);
const baseItemStyle = getStyle(8);

afterEach(cleanup);
mountTest(<Gap />);

describe('Gap[Base]', () => {
  displayTest(Gap, 'ShineoutGap');
  snapshotTest(<GapBase />);
  snapshotTest(<GapCustom />, 'about custom');
  snapshotTest(<GapVertical />, 'about vertical');
  snapshotTest(<GapAlign />, 'about align');
  snapshotTest(<GapSplit />, 'about split');
  // TODO: should test flex
  test('should render when set className and style', () => {
    const { container } = render(<Gap className='demo' style={{ width: 400 }} />);
    styleTest(container.querySelector('.demo')!, `${baseStyle} width: 400px;`);
  });
  test('should render default', () => {
    const { container } = render(<GapDemo className='demo' />);
    container
      .querySelector('.demo')
      ?.querySelectorAll('div')
      .forEach((item) => {
        styleTest(item, baseItemStyle);
      });
  });
  test('should render when set itemStyle', () => {
    const { container } = render(<GapDemo className='demo' itemStyle={{ width: 10 }} />);
    container
      .querySelector('.demo')
      ?.querySelectorAll('div')
      .forEach((item) => {
        styleTest(item, `width: 10px; ${baseItemStyle}`);
      });
  });
  test('should render when set column and row', () => {
    const { container } = render(<GapDemo className='demo' column={10} row={10} />);
    const wrapper = container.querySelector('.demo')!;
    styleTest(wrapper, `${getStyle(-10)} width: 400px;`);
    wrapper.querySelectorAll('div').forEach((item) => {
      styleTest(item, getStyle(10));
    });
  });
  test('should render when set vertical', () => {
    const { container } = render(
      <GapDemo className='demo' column={10} row={10} style={{ flexDirection: 'column' }} />,
    );
    styleTest(container.querySelector('.demo')!, `${getStyle(-10)} flex-direction: column;`);
  });
});
