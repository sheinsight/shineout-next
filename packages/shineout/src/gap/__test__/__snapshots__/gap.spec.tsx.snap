// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gap[Base] should render correctly  1`] = `
<div
  class="c-g3b8rb soui-gap-wrapper"
  style="margin-bottom: -8px; margin-right: -8px; width: 400px;"
>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 8px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
</div>
`;

exports[`Gap[Base] should render correctly about align 1`] = `
<div
  class="cuvzs0g soui-radio-group"
>
  <div
    class="ceo7z38 soui-radio-wrapper c875odf soui-radio-wrapper-checked"
  >
    <input
      checked=""
      style="display: none;"
      tabindex="0"
      type="radio"
    />
    <span
      class="chkciqb soui-radio-indicator-wrapper"
    >
      <span
        class="cp0qpb4 soui-radio-indicator"
        tabindex="1"
      />
    </span>
    <span
      class="c-v30ja8 soui-radio-desc"
    >
      flex-start
    </span>
  </div>
  <div
    class="ceo7z38 soui-radio-wrapper"
  >
    <input
      style="display: none;"
      tabindex="0"
      type="radio"
    />
    <span
      class="chkciqb soui-radio-indicator-wrapper"
    >
      <span
        class="cp0qpb4 soui-radio-indicator"
        tabindex="1"
      />
    </span>
    <span
      class="c-v30ja8 soui-radio-desc"
    >
      center
    </span>
  </div>
  <div
    class="ceo7z38 soui-radio-wrapper"
  >
    <input
      style="display: none;"
      tabindex="0"
      type="radio"
    />
    <span
      class="chkciqb soui-radio-indicator-wrapper"
    >
      <span
        class="cp0qpb4 soui-radio-indicator"
        tabindex="1"
      />
    </span>
    <span
      class="c-v30ja8 soui-radio-desc"
    >
      flex-end
    </span>
  </div>
  <div
    class="ceo7z38 soui-radio-wrapper"
  >
    <input
      style="display: none;"
      tabindex="0"
      type="radio"
    />
    <span
      class="chkciqb soui-radio-indicator-wrapper"
    >
      <span
        class="cp0qpb4 soui-radio-indicator"
        tabindex="1"
      />
    </span>
    <span
      class="c-v30ja8 soui-radio-desc"
    >
      baseline
    </span>
  </div>
</div>
`;

exports[`Gap[Base] should render correctly about custom 1`] = `
<div
  style="display: flex; align-items: center; margin-bottom: 16px;"
>
  <div
    style="width: 80px;"
  >
    column:
  </div>
  <div
    class="c7ubtly soui-slider-wrapper"
    style="flex: 1;"
  >
    <div
      class="c-l913oi soui-slider-track"
    >
      <div
        class="c-rukz6w soui-slider-track-inner"
        dir="ltr"
        style="left: 0%; right: 92%;"
      >
        <div
          class="c-446j6m soui-slider-indicator c-oxosmf soui-slider-indicator-end"
          dir="ltr"
        />
        <div
          class="clrhw2r soui-slider-end-value c-l884os soui-slider-value"
          dir="ltr"
        >
          8
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Gap[Base] should render correctly about vertical 1`] = `
<div
  class="c-g3b8rb soui-gap-wrapper"
  style="margin-bottom: -10px; margin-right: -8px; width: 400px; flex-direction: column;"
>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 10px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 10px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 10px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
  <div
    class="c-qchimb soui-gap-item"
    style="margin-bottom: 10px; margin-right: 8px;"
  >
    <button
      class="cq4fg10 soui-button-primary c-abf4z4 soui-button-button"
      dir="ltr"
      type="button"
    >
      <span>
        Button
      </span>
    </button>
  </div>
</div>
`;
